Table of content:

1. Variables
2. Functions
3. Strings
4. Lists
5. Dictionary
6. Tuples
7. Booleans
8. Functions again
9. Loops : for
10. If statement
11. While statement
12. Libraries import
13. Numpy
14. Pandas

## Python Tutorial


Welcome to this Python tutorial. The Jupyter Notebook interface is composed of several cells, in which you can write some code. When you run a cell, the only lines executed are those in the cell. However, they are linked by a single kernel : it means that you can define a variable in a cell and use it in an other.

## 1) Variables

To run a cell, select it and press "execute", or shift + enter. Try to run the following cell :

In [1]:
print("Congratulations")

Congratulations


If your code includes some functions which should display some informations, this information will appear under the box, like with "Congratulations" in the previous example.

#### &#x270d; 1.1) Display the number 13 and the message "Python is great"

In [2]:
print("13")
print("Python is great")

13
Python is great


#### &#x270d; 1.2) Create the variables $a = 5$, $b = 5.2$ and $c = 3 + 2j$

In [3]:
a, b = 5, 5.2
c = 3 + 2j

#### &#x270d; 1.3) Display $a$, $b$ and $c$ with the print function.  
They can be displayed with only one call by separating them with commas in the brackets.

In [4]:
print(a)
print(b)
print(c)

5
5.2
(3+2j)


#### &#x270d; 1.4) Display $a + b$ and $b * c$ without saving the product in a variable.

In [5]:
print(a + b)
print(b * c)

10.2
(15.600000000000001+10.4j)


#### &#x270d; 1.5) Check the type of $a$, $b$, $c$, then display and show the type of :  
$a + b$  
$a * b$  
$b - c$  
$a + b + c$  
$a * c$  
$c / a$  

In [8]:
print(type(a + b))
print(type(a*b))
print(type(b-c))
print(type(a + b + c))
print(type(a * c))
print(type( c / a))

<class 'float'>
<class 'float'>
<class 'complex'>
<class 'complex'>
<class 'complex'>
<class 'complex'>


## 2)Functions

#### &#x270d; 2.1) Save round(b) in a variable e, then display the type of :  
$b$  
$e$  
round

In [16]:
e = round(b)
print(type(b))
print(type(e))
print(type(round))

<class 'float'>
<class 'int'>
<class 'builtin_function_or_method'>


#### &#x270d; 2.2) Create a function which takes as argument a value $x$, and returns $x + x^2$ (show how your function works with example)

In [17]:
def exponent_x(x):
    return x + x**2

exponent_x(2)

6

#### &#x270d; 2.3) Create a function which takes as argument a value $x$, prints $x$, $2x$, $3x$, and returns $x - 1$. (show how your function works with example)

In [18]:
def product_x(x):
    print(x)
    print(2*x)
    print(3*x)
    return x - 1

product_x(3)

3
6
9


2

#### &#x270d; 2.4) Write down a function that takes $x$ and $y$ as the arguments, and returns a rounded value of $x$ multiplied by the value of $y$. (show how your function works with example)

In [19]:
def rounded_x (x, y):
    return round(x) * y

rounded_x(4.3, 2)

8

## 3) Strings

#### &#x270d; 3.1) Create a string called birthday containing your birthday.

In [20]:
birthday = "05.11"

#### &#x270d; 3.2) Using a and b, create a string "Hello my dear" and print it.

In [22]:
a = "Hello"
b = "my dear"
print(a, b)

Hello my dear


#### &#x270d; 3.3) Print the number of letters of a and b using the function len(string)

In [23]:
print(len(a + b))

12


A string can be included in the middle of another one using `%s`

Example: 

`animal = "beef"`

`action = "love"`

`dream = "i would %s to eat a %s" %(action,animal)`

`--> dream = "i would love to eat a beef"`

#### &#x270d; 3.4) Create the string "I am 30 years old and I live in Bogota" using the following strings :

In [26]:
age = 30
town = "Bogota"
presentation = "I am %i years old and I live in %s" %(age, town)
print(presentation)

I am 30 years old and I live in Bogota


## 4) Lists

To create a list, you need to use `[ ]`. If you're using a mac, the command is alt + shift + `(` or `)` .


#### &#x270d; 4.1) Create a list containing every letters from A to H, called letters.

In [80]:
letters = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H']

#### &#x270d; 4.2) Print the letter E using its index in the list letters.

Remember : index of the first element is $0$.

In [81]:
print(letters[4])

E


#### &#x270d; 4.3) Replace the letter G by Y in letters.

In [82]:
letters[6] = 'Y'

#### &#x270d; 4.4) Print the letter D in two different ways : once with its index, one with its index starting from the last element (negative index).

In [83]:
print(letters[3])
print(letters[-5])

D
D


#### &#x270d; 4.5) Add the letter P to the list letters, using the function letters.append()

In [84]:
letters.append('P')

#### &#x270d; 4.6) Create a sublist sublet = ["A","B","C"] from the list letters.

Remember : list[i:j] creates a sublist from list starting from the i th element to the j-1 th element.

In [85]:
sublet = letters[0:3]
print(sublet) 

['A', 'B', 'C']


#### &#x270d; 4.7) Concatenate letters and sublet into letters.

In [86]:
letters = letters + sublet

#### &#x270d; 4.8)  Erase every occurence of C in letters.

Remember : use list.remove(string) to delete every occurence of string in list.

In [91]:
letters.remove('C')

#### &#x270d; 4.9) Erase the second occurence of A in letters.
Remember : use list.pop(index) to delete the index th element.

In [93]:
print(letters)
letters.pop(-2)
print(letters)

['A', 'B', 'D', 'E', 'F', 'Y', 'H', 'P', 'A', 'B']
['A', 'B', 'D', 'E', 'F', 'Y', 'H', 'P', 'B']


## 5) Dictionary

Dictionaries are similiar to lists, but instead of accessing an element through its index, we can use some keywords, which can be of any type.
When you create it, you indicate the keyword and then the associated variables.
Try to print every variable of the following dict :

In [100]:
dict = {'x' : 3, 'y' : True, 'z' : [False, 89], 5 : "banana",  True : False }

#### &#x270d; 5.1) A list of the keywords of dict can be accessed using the function `dict.keys()`, use it.

In [101]:
print(dict.keys())

dict_keys(['x', 'y', 'z', 5, True])


#### &#x270d; 5.2) Add a new value False to dict, indexed by the keyword Banana.


To add a new keyword, give the associated value as if the keyword already existed. Example:  

`dict = {}`

`dict[3] = "hi"`

`--> dict = {3 : "hi"}`

In [102]:
dict['Banana'] = False
print(dict)

{'x': 3, 'y': True, 'z': [False, 89], 5: 'banana', True: False, 'Banana': False}


#### &#x270d; 5.3) Delete the key `y` from dict.

You can use `dict.pop(key)` to do that.

In [103]:
dict.pop('y')
print(dict)

{'x': 3, 'z': [False, 89], 5: 'banana', True: False, 'Banana': False}


#### &#x270d; 5.4) Try to delete the key `g`, which is not in dict.

In [104]:
dict.pop('g')

KeyError: 'g'

As you can see, instead of just ignoring the command, an error is returned. If you do not want the program to be stopped if the value is not existing, you can add another action to be performed if the key does not exist. It must be written as second parameter.

Addding None as second parameter prevents the program from being stopped by an unexisting key.

#### &#x270d; 5.5) Check this with the previous example.

In [105]:
dict.pop('g', None)

## 6) Tuples
Tuples are very similar to lists, except their shape cannot be changed.

#### &#x270d; 6.1) Create a tuple containing the integers from 5 to 9 and a tuple containing the integers from 1 to 4. Use ( ) to create them.

In [106]:
tuple_9 = (5, 6, 7, 8, 9)
tuple_4 = (1, 2, 3, 4)

#### &#x270d; 6.2) Try to sum up these tuples.

In [110]:
sum_tuple = tuple_9 + tuple_4
print(sum_tuple)

(5, 6, 7, 8, 9, 1, 2, 3, 4)


#### &#x270d; 6.3) Print the 5th element of the sum, using two different ways.

In [112]:
print(sum_tuple[5])

1


## 7) Booleans


#### &#x270d; 7.1) Create two booleans `a = True` and `b = False`, and check the result of every combination of `a` and `b` with the logical operators `and` / `or`.

In [113]:
a = True 
b = False
print(a and b)

False


#### &#x270d; 7.2) To invert a boolean, use the function not. Check the result of "not a or b"

In [114]:
print(not(a or b))

False


#### &#x270d; 7.3) Check how to use the different logical operators :  
equals ==  
less than <  
more than >    
less or equal <=
more or equal >=



In [115]:
print( a == b)
print(a > b)
print(a < b)
print(a <= b)
print(b >= b)

False
True
False
False
True


#### &#x270d; 7.4) Create a the boolean which indicates if 3 in in the list "list" or not.

Use the function in :  `a in list = True` if a is in list and `False` otherwise.

In [117]:
x = 3
number = [2,5,3]
a = x in number
print(a)

True


## 8) Functions again


Once created, a function is used only if it is called.

Here, the function square returns the squared value of the input.

#### &#x270d; 8.1) Try to print the squared value of 5 using the square function below.

In [118]:
def square(n):
    result = n**2
    return(result)

print(square(5))


25


The "return" word is very important when you define a function, for two reasons. First, it allows to tell what will be the output of the function. Second, this word means that we are at the end of the function, so the next lines do not need to be read.

#### &#x270d; 8.2) Create a function `square2(n)` identical to `square(n)` without the `return(result)` line and see what happens if you try to compute `square2(5)`.

In [120]:
def square2(n):
    result = n**2
print(square2(5))


None


#### &#x270d; 8.3) Create a function `square3(n)` identical to `square(n)` with one more line `print("hi")` after `return(result)` and see what happens if you try to compute `square3(5)`.

In [122]:
def square3(n):
    result = n**2
    return(result)
    print("hi")
print(square3(5))

25


#### &#x270d; 8.4) Create a function `firstandlast(l)` which returns a tuple containing the first and the last elements of `l`, a list input.
Try it on the list test.

In [123]:
def firstandlast(l): 
    return tuple(l)

tup = firstandlast([1, 3, 5, 8, 0])
print(tup)

(1, 3, 5, 8, 0)


Once a function is created, it can be called even in the definition of an other function.

#### &#x270d; 8.5) Create a function `sumoffandl` which takes as input a list, and returns the sum of its first and last terms. Use the `firstandlast(n)` function.

In [125]:
def firstandlast(n): 
    return [n[0], n[-1]]
            
def sumoffandl(l):
    return sum(firstandlast(l))

l = [1, 5, 7, 8]
sumoffandl(l)

9

## 9) Loops : `for`


Run the following code and watch the syntaxe of the For loop function

In [127]:
my_list = [2,5,4,2,8,20]
for x in my_list:
    print(x)

2
5
4
2
8
20


#### &#x270d; 9.1) Now try your own loop : compute the sum of the terms of the list `my_list` with a for loop.

In [128]:
sum_num = 0
for i in my_list: 
    sum_num += i
    
print(sum_num)
    

41


When we want to iterate on some integers, we use the range function : 

`for k in range(i,j)`

 means that `k` will take the value of every integer between `i` and `j-1`

#### &#x270d; 9.2) Compute the product of integers from 37 to 54 using a for loop and the range function.

In [129]:
for i in range(37, 55): 
    print(i)

37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54


#### &#x270d; 9.3) Create a function which takes a list as input and returns the same list but elements are inverted (the first element becomes the last, etc)

In [20]:
def inversion(l):
    return list(reversed(l))

liste = [1, 5, 6, 3]
inversion(liste)


[3, 6, 5, 1]

## 10) `If` statement

#### &#x270d; 10.1) Test the function test with -5, 1, -1.7, 8, and interpret.

In [24]:
def test(x):
    if x**2 > 4 : #true/false
        print("yes")
    else :
        print("no")

test(-5)
test(1)
test(-1.7)
test(8)


yes
no
no
yes


#### &#x270d; 10.2) Create a function which takes a list as input and returns a boolean, set to `True` if the first element is a string, an `False` otherwise.

In [46]:
def verify(l):
    element = l[0]
    if isinstance(element, str): 
        return True
    else :
        return False
    
l = [4, 5, 4, "low"]
verify(l)

False

#### &#x270d; 10.3) Create a function which takes a list as input and returns a boolean, set to `True` if there is a string in the list, and `False` otherwise.

Hint : combine `for` and `if`.

In [54]:

def verifying(l):
    op = 0
    for i in l : 
        if isinstance(i, str):
            op = 1 
            return True
        else:
            continue 
    if op == 0: 
        return False
 
l = [4, 5, 4, 'sd']
verifying(l)

True

## 11) `While` statement

#### &#x270d; 11.1) Run the following cell and interpret.

In [55]:
x = 16
while x > 3 :
    print(x)
    x = x / 2

16
8.0
4.0


#### &#x270d; 11.2) Create a function which takes a list, and returns false if there is no string inside, and the index of the first string otherwise, using a while loop.

In [56]:
def index_first(l):
    i = 0
    while i < len(l):
        if isinstance(l[i], str):
            return i
        i += 1
    return False
l = [4, 5, "h", 2]
index_first(l)

2

#### &#x270d; 11.3) Print the first integer whose squared value is more than 7533493

In [62]:
number = 1
while number**2 <= 7533493:
    number += 1
print(number)


2745


## 12) Libraries import

### &#x270d; 12.1) Import the library numpy, with the shortened name np.

Remember : `import XXXXXX as YY` to import the module `XXXXXX` with the shortened name `YY`.

In [57]:
import numpy as np 

Once a library is imported, all its functions are stored in the kernel and can be accessed.

#### &#x270d; 12.2) Print the value of pi using the function pi in numpy.

Remember : To use a function of a library imported as YY, use YY.function()

In [63]:
print(np.pi)

3.141592653589793


## 13) Numpy

#### &#x270d; 13.1) Create an array containing the integers from 0 to 9

In [64]:
np_array = np.array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
print(np_array)


[0 1 2 3 4 5 6 7 8 9]


#### &#x270d; 13.2) Create a 10x10 matrix name M, use arange and reshape

In [65]:
M = np.arange(100).reshape((10, 10))
print(M)


[[ 0  1  2  3  4  5  6  7  8  9]
 [10 11 12 13 14 15 16 17 18 19]
 [20 21 22 23 24 25 26 27 28 29]
 [30 31 32 33 34 35 36 37 38 39]
 [40 41 42 43 44 45 46 47 48 49]
 [50 51 52 53 54 55 56 57 58 59]
 [60 61 62 63 64 65 66 67 68 69]
 [70 71 72 73 74 75 76 77 78 79]
 [80 81 82 83 84 85 86 87 88 89]
 [90 91 92 93 94 95 96 97 98 99]]


#### &#x270d; 13.4) Compute the sum of the lists X and Y without using the function for.

In [67]:
X = [8, 2, 3, 2, 5]
Y = [6, 9, -1, 9, 1]


Z = [x + y for x, y in zip(X, Y)]
print(Z)

Z = list(map(lambda x, y: x + y, X, Y))
print(Z)


X = np.array(X)
Y = np.array(Y)
Z = X + Y
print(Z)


[14, 11, 2, 11, 6]
[14, 11, 2, 11, 6]
[14 11  2 11  6]


#### &#x270d; 13.5) Compute the scalar product of X and Y.

Hint : the numpy function is `dot`.

A square matrix, in which all diagonal elements are equal to same scalar and all other elements are zero, is called a scalar matrix. Or. A diagonal matrix, in which all diagonal elements are equal to same scalar, is called a scalar matrix. A.

In [70]:
X = [1, 0, 3]
Y = [1, 5, 6]

# To compute the scalar product of X and Y
result = np.dot(X, Y)
print(result)


19


## 14) Pandas

In Pandas, we work with dataframes. Let's create one to see how it works.

#### &#x270d; 14.1) Import the pandas module, with pd as shortened name, and the random module. 

In [69]:
import pandas as pd

#### &#x270d; 14.2) Create 5 lists composed of 4 random numbers between 1 and 100 and a string as final element.

Use random.randint(i,n) from numpy to get a list of n random numbers between 0 and i-1.

In [71]:
# Create 5 lists of 4 random numbers and a string
lists = []
for i in range(5):
    rand_nums = list(np.random.randint(1, 101, size=4))
    rand_str = 'string_' + str(i+1)
    rand_nums.append(rand_str)
    lists.append(rand_nums)

#### &#x270d; 14.3) Create a dataframe named data, composed of these 5 lists. We will call the columns F1, F2, F3, F4 and STR.



In [73]:
# Create a DataFrame from the lists with column names
data = pd.DataFrame(lists, columns=['F1', 'F2', 'F3', 'F4', 'STR'])

# Print the DataFrame
print(data)



   F1  F2  F3  F4       STR
0  19  40  81  32  string_1
1  90  37  40  15  string_2
2  89  52  98  86  string_3
3  14  76  75  10  string_4
4  88  96  64  15  string_5


#### &#x270d; 14.4) Print the STR feature of each element.

In [74]:
# Print the STR feature of each element
for index, row in data.iterrows():
    print(row['STR'])

string_1
string_2
string_3
string_4
string_5


#### &#x270d; 14.5) Create a new dataframe composed of the rows 1, 3 and 4 of data. Compute its F3 mean.



In [76]:
# Select rows 1, 3, and 4 of data to create a new DataFrame
new_data = data.iloc[[1, 3, 4]]

# Compute the mean of the F3 column of the new DataFrame
mean_f3 = new_data['F3'].mean()

print(mean_f3)

59.666666666666664


#### &#x270d; 14.6) Let's go back to the dataframe `data`. Create a new dataframe composed of the daqta rows where F1 > 40.



In [77]:
# Create a new DataFrame composed of rows where F1 > 40
new_data = data[data['F1'] > 40]

# Print the new DataFrame
print(new_data)

   F1  F2  F3  F4       STR
1  90  37  40  15  string_2
2  89  52  98  86  string_3
4  88  96  64  15  string_5
